File: browser/src/engine/strategies.ts
// Browser port of server/strategies.py. Each strategy is a human-replicable
// rule reasoning at most one move ahead; ties are broken at random.
import { divmod } from '../utils';
import {
  applyToSets,
  EngineState,
  GOAT_PLAYER,
  JUMP_NEIGHBOURS,
  Move,
  possibleMoves,
  STEP_NEIGHBOURS,
  TIGER_PLAYER,
} from './game';

export const STRONG_POINTS = new Set(
  [...Array(25).keys()].filter((n) => STEP_NEIGHBOURS[n].length === 8),
);

function isEdge(pos: number): boolean {
  const [y, x] = divmod(pos, 5);
  return x === 0 || x === 4 || y === 0 || y === 4;
}

export const EDGE_POINTS = new Set([...Array(25).keys()].filter(isEdge));

function manhattanToCentre(pos: number): number {
  const [y, x] = divmod(pos, 5);
  return Math.abs(x - 2) + Math.abs(y - 2);
}

type Sets = [Set<number>, Set<number>]; // [tigers, goats]

function applied(move: Move, tigers: Set<number>, goats: Set<number>): Sets {
  const t = new Set(tigers);
  const g = new Set(goats);
  applyToSets(move, t, g);
  return [t, g];
}

function capturesOf(tigers: Set<number>, goats: Set<number>): Move[] {
  const occ = new Set([...tigers, ...goats]);
  const moves: Move[] = [];
  for (const t of tigers) {
    for (const dest of JUMP_NEIGHBOURS[t]) {
      const eaten = (t + dest) / 2;
      if (goats.has(eaten) && !occ.has(dest)) {
        moves.push([t, eaten, dest]);
      }
    }
  }
  return moves;
}

function mobilityOf(tigers: Set<number>, goats: Set<number>): number {
  const occ = new Set([...tigers, ...goats]);
  let steps = 0;
  for (const t of tigers) {
    for (const dest of STEP_NEIGHBOURS[t]) {
      if (!occ.has(dest)) {
        steps += 1;
      }
    }
  }
  return capturesOf(tigers, goats).length + steps;
}

const numCapturable = (tigers: Set<number>, goats: Set<number>): number =>
  capturesOf(tigers, goats).length;

const goatDest = (move: Move): number => (move.length === 1 ? move[0] : move[1]);

// Compare composite keys (booleans count as 1/0), highest wins.
function cmp(a: number[], b: number[]): number {
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      return a[i] - b[i];
    }
  }
  return 0;
}

const randChoice = <T>(items: T[]): T =>
  items[Math.floor(Math.random() * items.length)];

function argbest(items: Move[], keyFn: (m: Move) => number[]): Move {
  let best: number[] | null = null;
  let bucket: Move[] = [];
  for (const item of items) {
    const k = keyFn(item);
    if (best === null || cmp(k, best) > 0) {
      best = k;
      bucket = [item];
    } else if (cmp(k, best) === 0) {
      bucket.push(item);
    }
  }
  return randChoice(bucket);
}

export interface Strategy {
  name: string;
  side: number;
  description: string;
  choose(state: EngineState): Move;
}

// --- goat strategies ---------------------------------------------------------

function safeGoatMoves(state: EngineState, moves: Move[]): Move[] {
  const { tigers, goats } = state;
  const before = numCapturable(tigers, goats);
  const safe = moves.filter((m) => {
    const [t, g] = applied(m, tigers, goats);
    return numCapturable(t, g) <= before;
  });
  return safe.length ? safe : moves;
}

function goatAdjacency(move: Move, goats: Set<number>): number {
  const dest = goatDest(move);
  const src = move.length === 2 ? move[0] : null;
  return STEP_NEIGHBOURS[dest].filter((nb) => goats.has(nb) && nb !== src).length;
}

const goatStrategy = (
  name: string,
  description: string,
  prefer: (state: EngineState, moves: Move[]) => Move,
): Strategy => ({
  name,
  side: GOAT_PLAYER,
  description,
  choose(state) {
    return prefer(state, safeGoatMoves(state, possibleMoves(state)));
  },
});

export const GOAT_STRATEGIES: Strategy[] = [
  {
    name: 'goat-random',
    side: GOAT_PLAYER,
    description: 'Play a random legal move (placement or step). Baseline.',
    choose: (state) => randChoice(possibleMoves(state)),
  },
  goatStrategy(
    'goat-safe',
    'Never put a goat where a tiger could jump it next turn; otherwise pick ' +
      'freely among the safe moves.',
    (_state, moves) => randChoice(moves),
  ),
  goatStrategy(
    'goat-safe-block',
    'Among safe moves, choose the one that leaves the tigers with the fewest ' +
      'moves — hem the tigers in while never offering a capture.',
    (state, moves) =>
      argbest(moves, (m) => [-mobilityOf(...applied(m, state.tigers, state.goats))]),
  ),
  goatStrategy(
    'goat-safe-connected',
    'Among safe moves, keep goats touching each other — pick the move whose ' +
      'goat ends up next to the most friendly goats, building a solid wall.',
    (state, moves) => argbest(moves, (m) => [goatAdjacency(m, state.goats)]),
  ),
  goatStrategy(
    'goat-safe-strong-points',
    'Among safe moves, grab the strong central points first (the most ' +
      'connected squares); break ties by hemming the tigers in.',
    (state, moves) =>
      argbest(moves, (m) => [
        STRONG_POINTS.has(goatDest(m)) ? 1 : 0,
        -mobilityOf(...applied(m, state.tigers, state.goats)),
      ]),
  ),
  goatStrategy(
    'goat-safe-edge',
    'Among safe moves, build from the rim — prefer perimeter squares, which ' +
      'give tigers fewer places to land; break ties by staying connected.',
    (state, moves) =>
      argbest(moves, (m) => [
        EDGE_POINTS.has(goatDest(m)) ? 1 : 0,
        goatAdjacency(m, state.goats),
      ]),
  ),
];

// --- tiger strategies --------------------------------------------------------

const split = (moves: Move[]): [Move[], Move[]] => [
  moves.filter((m) => m.length === 3),
  moves.filter((m) => m.length === 2),
];

const bestCapture = (state: EngineState, captures: Move[]): Move =>
  argbest(captures, (m) => [mobilityOf(...applied(m, state.tigers, state.goats))]);

const tigerStrategy = (
  name: string,
  description: string,
  pick: (state: EngineState, captures: Move[], steps: Move[]) => Move,
): Strategy => ({
  name,
  side: TIGER_PLAYER,
  description,
  choose(state) {
    const [captures, steps] = split(possibleMoves(state));
    return pick(state, captures, steps);
  },
});

export const TIGER_STRATEGIES: Strategy[] = [
  tigerStrategy(
    'tiger-random',
    'Play a random legal move, capture or not. Baseline.',
    (_state, captures, steps) => randChoice([...captures, ...steps]),
  ),
  tigerStrategy(
    'tiger-greedy-capture',
    'Always take a capture when one exists (any of them); otherwise step at ' +
      "random. The simple 'eat whenever you can' instinct.",
    (_state, captures, steps) =>
      captures.length ? randChoice(captures) : randChoice(steps),
  ),
  tigerStrategy(
    'tiger-mobile-capture',
    'Always capture when possible, but pick the capture (and otherwise the ' +
      'step) that keeps the tigers with the most moves — eat without trapping ' +
      'yourself.',
    (state, captures, steps) =>
      captures.length
        ? bestCapture(state, captures)
        : argbest(steps, (m) => [mobilityOf(...applied(m, state.tigers, state.goats))]),
  ),
  tigerStrategy(
    'tiger-threat-builder',
    'Capture when possible; otherwise move to threaten the most goats next ' +
      "turn, setting up jumps (including forks the goat can't fully answer).",
    (state, captures, steps) =>
      captures.length
        ? bestCapture(state, captures)
        : argbest(steps, (m) => {
            const [t, g] = applied(m, state.tigers, state.goats);
            return [numCapturable(t, g), mobilityOf(t, g)];
          }),
  ),
  tigerStrategy(
    'tiger-centre-control',
    'Capture when possible; otherwise march tigers toward the strong central ' +
      "points, controlling the board's busiest lines.",
    (state, captures, steps) =>
      captures.length
        ? bestCapture(state, captures)
        : argbest(steps, (m) => [
            STRONG_POINTS.has(m[1]) ? 1 : 0,
            -manhattanToCentre(m[1]),
          ]),
  ),
];

export const STRATEGY_BY_NAME: Record<string, Strategy> = Object.fromEntries(
  [...GOAT_STRATEGIES, ...TIGER_STRATEGIES].map((s) => [s.name, s]),
);

File: server/strategies.py
"""Human-understandable, human-replicable playing strategies for Bagh-Chal.

Every strategy here is a rule a person could follow at the board: look at the
position, the legal moves, and reason at most one move ahead (e.g. "can this
goat be jumped next turn?", "does this keep my tigers mobile?"). No strategy
uses game-tree search or opaque evaluation weights — that is a hard constraint,
so the strategies stay explainable and reproducible by a human.

Choosing among equally-good moves uses a seeded RNG so games are reproducible.
"""

from .constants import GOAT_PLAYER, TIGER_PLAYER
from .graph import JUMPS_GRAPH, STEPS_GRAPH
from .notations import pos_num_to_coord

# The five "strong points": board points with the most connections (eight
# lines each). Controlling them is a classic human heuristic.
STRONG_POINTS = frozenset(n for n in range(25) if len(STEPS_GRAPH[n]) == 8)


def _is_edge(pos):
    x, y = pos_num_to_coord(pos)
    return x in (0, 4) or y in (0, 4)


# Perimeter points (any point on the outer ring). Fewer lines run through them,
# which limits the landing squares a tiger needs to jump.
EDGE_POINTS = frozenset(n for n in range(25) if _is_edge(n))

CENTRE = pos_num_to_coord(12)


def _manhattan_to_centre(pos):
    x, y = pos_num_to_coord(pos)
    return abs(x - CENTRE[0]) + abs(y - CENTRE[1])


# --- pure position helpers (mirror server/game.py move rules) -----------------


def tiger_captures(tigers, goats):
    """All immediate tiger jump-captures as (src, eaten, dest) tuples."""
    occupied = tigers | goats
    moves = []
    for t in tigers:
        for d in JUMPS_GRAPH[t]:
            eaten = (t + d) // 2
            if eaten in goats and d not in occupied:
                moves.append((t, eaten, d))
    return moves


def tiger_steps(tigers, goats):
    """All immediate tiger single steps as (src, dest) tuples."""
    occupied = tigers | goats
    moves = []
    for t in tigers:
        for d in STEPS_GRAPH[t]:
            if d not in occupied:
                moves.append((t, d))
    return moves


def tiger_mobility(tigers, goats):
    return len(tiger_captures(tigers, goats)) + len(tiger_steps(tigers, goats))


def num_capturable(tigers, goats):
    """How many capture moves the tiger side has right now."""
    return len(tiger_captures(tigers, goats))


def apply_to_sets(move, tigers, goats):
    """Return (tigers, goats) sets after applying a move tuple."""
    t, g = set(tigers), set(goats)
    if len(move) == 1:
        g.add(move[0])
    elif len(move) == 2:
        src, dest = move
        if src in t:
            t.discard(src)
            t.add(dest)
        else:
            g.discard(src)
            g.add(dest)
    else:
        src, eaten, dest = move
        t.discard(src)
        t.add(dest)
        g.discard(eaten)
    return t, g


def goat_dest(move):
    """The square a goat ends up on for a placement (len 1) or step (len 2)."""
    return move[0] if len(move) == 1 else move[1]


# --- strategy base ------------------------------------------------------------


class Strategy:
    name = "base"
    side = None  # GOAT_PLAYER or TIGER_PLAYER
    description = ""

    def choose(self, game, rng):  # pragma: no cover - overridden
        raise NotImplementedError

    def __repr__(self):
        return self.name


def _argbest(items, key, rng):
    """Pick the item with the max key, breaking ties with the seeded rng."""
    best = None
    best_keys = []
    for item in items:
        k = key(item)
        if best is None or k > best:
            best = k
            best_keys = [item]
        elif k == best:
            best_keys.append(item)
    return rng.choice(best_keys)


# --- goat strategies ----------------------------------------------------------


class GoatStrategy(Strategy):
    side = GOAT_PLAYER

    def safe_moves(self, game, moves):
        """Moves that do not hand the tiger a new capture next turn."""
        tigers, goats = game.pieces.tigers, game.pieces.goats
        before = num_capturable(tigers, goats)
        safe = []
        for move in moves:
            nt, ng = apply_to_sets(move, tigers, goats)
            if num_capturable(nt, ng) <= before:
                safe.append(move)
        return safe or list(moves)

    def prefer(self, game, moves, rng):  # default: random among the pool
        return rng.choice(moves)

    def choose(self, game, rng):
        moves = game.possible_moves()
        return self.prefer(game, self.safe_moves(game, moves), rng)


class GoatRandom(GoatStrategy):
    name = "goat-random"
    description = "Play a random legal move (placement or step). Baseline."

    def safe_moves(self, game, moves):  # ignores safety
        return list(moves)


class GoatSafe(GoatStrategy):
    name = "goat-safe"
    description = (
        "Never put a goat where a tiger could jump it next turn; otherwise pick "
        "freely among the safe moves."
    )


class GoatSafeBlock(GoatStrategy):
    name = "goat-safe-block"
    description = (
        "Among safe moves, choose the one that leaves the tigers with the fewest "
        "moves — hem the tigers in while never offering a capture."
    )

    def prefer(self, game, moves, rng):
        tigers, goats = game.pieces.tigers, game.pieces.goats
        return _argbest(
            moves,
            key=lambda m: -tiger_mobility(*apply_to_sets(m, tigers, goats)),
            rng=rng,
        )


class GoatSafeConnected(GoatStrategy):
    name = "goat-safe-connected"
    description = (
        "Among safe moves, keep goats touching each other — pick the move whose "
        "goat ends up next to the most friendly goats, building a solid wall."
    )

    def prefer(self, game, moves, rng):
        goats = game.pieces.goats

        def adjacency(m):
            dest = goat_dest(m)
            src = m[0] if len(m) == 2 else None
            return sum(1 for nb in STEPS_GRAPH[dest] if nb in goats and nb != src)

        return _argbest(moves, key=adjacency, rng=rng)


class GoatSafeStrong(GoatStrategy):
    name = "goat-safe-strong-points"
    description = (
        "Among safe moves, grab the strong central points first (the most "
        "connected squares); break ties by hemming the tigers in."
    )

    def prefer(self, game, moves, rng):
        tigers, goats = game.pieces.tigers, game.pieces.goats
        return _argbest(
            moves,
            key=lambda m: (
                goat_dest(m) in STRONG_POINTS,
                -tiger_mobility(*apply_to_sets(m, tigers, goats)),
            ),
            rng=rng,
        )


class GoatSafeEdge(GoatStrategy):
    name = "goat-safe-edge"
    description = (
        "Among safe moves, build from the rim — prefer perimeter squares, which "
        "give tigers fewer places to land; break ties by staying connected."
    )

    def prefer(self, game, moves, rng):
        goats = game.pieces.goats

        def key(m):
            dest = goat_dest(m)
            src = m[0] if len(m) == 2 else None
            adj = sum(1 for nb in STEPS_GRAPH[dest] if nb in goats and nb != src)
            return (dest in EDGE_POINTS, adj)

        return _argbest(moves, key=key, rng=rng)


# --- tiger strategies ---------------------------------------------------------


class TigerStrategy(Strategy):
    side = TIGER_PLAYER

    def choose(self, game, rng):
        moves = game.possible_moves()
        captures = [m for m in moves if len(m) == 3]
        steps = [m for m in moves if len(m) == 2]
        return self.pick(game, captures, steps, rng)

    def pick(self, game, captures, steps, rng):  # pragma: no cover - overridden
        raise NotImplementedError

    def best_capture(self, game, captures, rng):
        """Default capture choice: eat, keeping the most tiger mobility."""
        tigers, goats = game.pieces.tigers, game.pieces.goats
        return _argbest(
            captures,
            key=lambda m: tiger_mobility(*apply_to_sets(m, tigers, goats)),
            rng=rng,
        )


class TigerRandom(TigerStrategy):
    name = "tiger-random"
    description = "Play a random legal move, capture or not. Baseline."

    def pick(self, game, captures, steps, rng):
        return rng.choice(captures + steps)


class TigerGreedyCapture(TigerStrategy):
    name = "tiger-greedy-capture"
    description = (
        "Always take a capture when one exists (any of them); otherwise step at "
        "random. The simple 'eat whenever you can' instinct."
    )

    def pick(self, game, captures, steps, rng):
        if captures:
            return rng.choice(captures)
        return rng.choice(steps)


class TigerMobileCapture(TigerStrategy):
    name = "tiger-mobile-capture"
    description = (
        "Always capture when possible, but pick the capture (and otherwise the "
        "step) that keeps the tigers with the most moves — eat without trapping "
        "yourself."
    )

    def pick(self, game, captures, steps, rng):
        tigers, goats = game.pieces.tigers, game.pieces.goats
        if captures:
            return self.best_capture(game, captures, rng)
        return _argbest(
            steps,
            key=lambda m: tiger_mobility(*apply_to_sets(m, tigers, goats)),
            rng=rng,
        )


class TigerThreatBuilder(TigerStrategy):
    name = "tiger-threat-builder"
    description = (
        "Capture when possible; otherwise move to threaten the most goats next "
        "turn, setting up jumps (including forks the goat can't fully answer)."
    )

    def pick(self, game, captures, steps, rng):
        tigers, goats = game.pieces.tigers, game.pieces.goats
        if captures:
            return self.best_capture(game, captures, rng)
        return _argbest(
            steps,
            key=lambda m: (
                num_capturable(*apply_to_sets(m, tigers, goats)),
                tiger_mobility(*apply_to_sets(m, tigers, goats)),
            ),
            rng=rng,
        )


class TigerCentreControl(TigerStrategy):
    name = "tiger-centre-control"
    description = (
        "Capture when possible; otherwise march tigers toward the strong central "
        "points, controlling the board's busiest lines."
    )

    def pick(self, game, captures, steps, rng):
        if captures:
            return self.best_capture(game, captures, rng)
        return _argbest(
            steps,
            key=lambda m: (m[1] in STRONG_POINTS, -_manhattan_to_centre(m[1])),
            rng=rng,
        )


# --- registries ---------------------------------------------------------------

GOAT_STRATEGIES = [
    GoatRandom(),
    GoatSafe(),
    GoatSafeBlock(),
    GoatSafeConnected(),
    GoatSafeStrong(),
    GoatSafeEdge(),
]

TIGER_STRATEGIES = [
    TigerRandom(),
    TigerGreedyCapture(),
    TigerMobileCapture(),
    TigerThreatBuilder(),
    TigerCentreControl(),
]
